src/darwin/Framework/CHIP/MTROperationalCredentialsDelegate.mm

```diff
@@ -112,9 +115,171 @@
+
+    Callback::Callback<chip::Controller::OnNOCChainGeneration> * onCompletion = mOnNOCCompletionCallback;
+    mOnNOCCompletionCallback = nullptr;
+
+    // Call-back into commissioner with the generated data.
+    dispatch_sync(mChipWorkQueue, ^{
+        onCompletion->mCall(onCompletion->mContext, status, noc, icac, rcac, ipk, adminSubject);
+    });
+
+    return CHIP_NO_ERROR;
+}
+
 CHIP_ERROR MTROperationalCredentialsDelegate::GenerateNOCChain(const chip::ByteSpan & csrElements, const chip::ByteSpan & csrNonce,
     const chip::ByteSpan & attestationSignature, const chip::ByteSpan & attestationChallenge, const chip::ByteSpan & DAC,
     const chip::ByteSpan & PAI, chip::Callback::Callback<chip::Controller::OnNOCChainGeneration> * onCompletion)
+{
+    if (mNocChainIssuer != nil) {
+        return CallbackGenerateNOCChain(csrElements, csrNonce, attestationSignature, attestationChallenge, DAC, PAI, onCompletion);
+    } else {
+        return LocalGenerateNOCChain(csrElements, csrNonce, attestationSignature, attestationChallenge, DAC, PAI, onCompletion);
+    }
+}
+
+CHIP_ERROR MTROperationalCredentialsDelegate::CallbackGenerateNOCChain(const chip::ByteSpan & csrElements,
+    const chip::ByteSpan & csrNonce, const chip::ByteSpan & csrElementsSignature, const chip::ByteSpan & attestationChallenge,
+    const chip::ByteSpan & DAC, const chip::ByteSpan & PAI,
+    chip::Callback::Callback<chip::Controller::OnNOCChainGeneration> * onCompletion)
+{
+    mOnNOCCompletionCallback = onCompletion;
+
+    TLVReader reader;
+    reader.Init(csrElements);
+
+    if (reader.GetType() == kTLVType_NotSpecified) {
+        ReturnErrorOnFailure(reader.Next());
+    }
+
+    VerifyOrReturnError(reader.GetType() == kTLVType_Structure, CHIP_ERROR_WRONG_TLV_TYPE);
+    VerifyOrReturnError(reader.GetTag() == AnonymousTag(), CHIP_ERROR_UNEXPECTED_TLV_ELEMENT);
+
+    TLVType containerType;
+    ReturnErrorOnFailure(reader.EnterContainer(containerType));
+    ReturnErrorOnFailure(reader.Next(kTLVType_ByteString, TLV::ContextTag(1)));
+
+    chip::ByteSpan csr;
+    reader.Get(csr);
+    reader.ExitContainer(containerType);
+
+    CSRInfo * csrInfo = [[CSRInfo alloc] initWithNonce:AsData(csrNonce)
+                                              elements:AsData(csrElements)
+                                     elementsSignature:AsData(csrElementsSignature)
+                                                   csr:AsData(csr)];
+
+    chip::ByteSpan certificationDeclarationSpan;
+    chip::ByteSpan attestationNonceSpan;
+    uint32_t timestampDeconstructed;
+    chip::ByteSpan firmwareInfoSpan;
+    chip::Credentials::DeviceAttestationVendorReservedDeconstructor vendorReserved;
+
+    __block chip::Optional<chip::Controller::CommissioningParameters> commissioningParameters;
+    // Dereferencing mCppCommissioner as it would be set to point to a valid Cpp commissioner by now, as we are in the middle of
+    // commissioning
+    dispatch_sync(mChipWorkQueue, ^{
```

Comment:
Isn't this _already_ running on the chip work queue?  In which case, won't this deadlock?